### DISCUSSION TITLE
Typst 'asis' output

### DISCUSSION THREAD

**Discussion by baptiste** [quarto-dev/quarto-cli#9163] [Q&A] [answered]
### Description

I'm trying to use a custom knitr engine to enter raw yaml in a chunk and output a raw typst string in a quarto document using the typst output format, but I'm a bit stumped with the output. It seems that the text is escaped, and I don't know how to turn that off.

MRE:

````
---
title: "Test"
format: 
  typst:
    keep-typ: true
    keep-md: true
---

```{r, echo=FALSE}
library(glue)
library(yaml)
library(knitr)

knitr::knit_engines$set(yamltypst = function(options) {

    options[["results"]] <- "asis"

    # get code together
    code <- paste(options$code, collapse = '\n')

    # parse
    spec <- yaml::yaml.load(code)
    
    # reset language for markdown highlighting
    options$engine <- "yaml"

    out <- glue::glue("#strong(  [{ spec }] )")
    
    knitr::engine_output(
      options,
      options$code,
      out = out
    )
  })

```

## Quarto

Quarto enables you to weave together content and executable code into a finished document. 

```{yamltypst, echo=FALSE, asis=TRUE}
stuff: yamlcontent
```

````

Output:
````
[..truncated..]

#block[
\#strong\( \[yamlcontent\] )

]

````

Output I am hoping to get (no block container, and no escaping of characters):

````

[..truncated..]

#strong( [yamlcontent] )
````

So that typst processes it. When I switch to a different format, eg. `format: markdown`, this quoting doesn't seem to happen. Can it be turned off?

**Comment by mcanouil** [comment]
If you want to add/emit raw "whatever":
- `` `something`{=whatever} ``
- ` ```{=whatever} `
FYI, the recommended syntax for code cell option in Quarto in YAML not R inline.

**Comment by cderv** [comment, marked as answer]
**knitr** has a function to emit raw block - you should use it. 
````r
out <- knitr::raw_block(glue::glue("#strong(  [{ spec }] )"), "typst")
````

Also in `` ```{yamltypst, echo=FALSE, asis=TRUE} `` , the `asis` option is not a known **knitr** options. 

You already used `options[["results"]] <- "asis"` in your engine, so it should be enough. 

Hope it helps

````markdown
---
title: "Test"
format: 
  typst:
    keep-typ: true
    keep-md: true
---

```{r, echo=FALSE}
library(glue)
library(yaml)
library(knitr)

knitr::knit_engines$set(yamltypst = function(options) {

    options[["results"]] <- "asis"

    # get code together
    code <- paste(options$code, collapse = '\n')

    # parse
    spec <- yaml::yaml.load(code)
    
    # reset language for markdown highlighting
    options$engine <- "yaml"

    out <- knitr::raw_block(glue::glue("#strong(  [{ spec }] )"), "typst")
    
    knitr::engine_output(
      options,
      options$code,
      out = out
    )
  })
```

## Quarto

Quarto enables you to weave together content and executable code into a finished document.

```{yamltypst}
stuff: yamlcontent
```

````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/70863a96-c1e5-4074-a91a-7b2ad2944d09)

**Reply by baptiste** [reply]
Perfect, thank you! I added the second 'asis' out of desperation..

**Reply by baptiste** [reply]
Is there a way to not have the `#block[]` container? It's preventing me from adding a `#set page()` directive, as it's not allowed within a block.

**Reply by cscheid** [reply]
What's happening here is that Pandoc adds a `#block[]` call around every div:

````
$ pandoc -f markdown -t typst
::: {}
Hello
:::
^D
#block[
Hello

]
````

I think Quarto is wrapping your command around a Div, and then (as it stands), I think you're out of luck. Can you give us a minimal example? We have mechanisms for stripping structural Divs out of documents, and we should consider using it here.

**Reply by baptiste** [reply]
Thanks, indeed that's what I feared :/ 

@cderv 's answer is a good example. Adding `#set page(flipped: true)` to the contents results in an error,

<img width="776" alt="Screenshot 2024-03-25 at 07 18 27" src="https://github.com/quarto-dev/quarto-cli/assets/18970/64766463-6d8f-4a6f-afd1-f25d778faf80">

**Reply by jimjam-slam** [reply]
Not sure if this would work for your needs, @baptiste, but I had my R code emit typist raw blocks, and that seemed to work okay (although in my case, I did the set rules in an actual Typst raw block before the R code).

This is from a calendar I made to kick the tyres on Typst. It ingests a spreadsheet of events and outputs a a two-column grid layout that looks like this (excuse the redacted stuff; this was an internal test).

![Screenshot 2024-03-25 at 10 07 25](https://github.com/quarto-dev/quarto-cli/assets/6520659/51849fb7-9a47-4c79-8b00-916a6353cad0)

The specifics of the generating code aren't important, but I can confirm that neither the literal raw Typst block nor the one emitted by the R chunk below are wrapped in a div (see redacted Typst intermediary file at the bottom).

````

```{r}
# some data processing goes here
```

```{=typst}
#set text(font: "Roboto")

#set page(
  header: align(right)[
    #image("logo.png", width: 15%) 
  ]
)

#let cell = rect.with(
  inset: 0pt,
  stroke: white,
  width: 100%,
  radius: 6pt)

#let fa(codepoint) = text(font: "Font Awesome 6 Free", codepoint)
#let calendar = fa("\u{f133}")

#let deemph(x) = text(fill: rgb("1978b0"), x)

= My title here \
As of #datetime.today().display("[day] [month repr:long] [year]")
  \
  \
```

```{r}
#| label: output
#| echo: false
#| output: asis

cat("```{=typst}\n")

grid_start = 
  "#grid(
  columns: 2,
  rows: (100pt, 100pt, 100pt, 100pt),
  column-gutter: 15pt,\n\n"
grid_end = ")\n#pagebreak(weak: true)"

story_cell <- function(publish, content, event) {
  cell = paste0(
    "cell(height: 100%)[ \\\n",
    day(publish), " ", month.abb[month(publish)], "\\\n",
    "*", content, "*"
  )
  if (!is.na(event)) {
    cell = paste0(
      cell, "\\\n",
      "#deemph[",
        "#calendar ",
        str_trunc(hook, width = 80),
      "]")
  }
  cell = paste0(cell, " \\\n]")

  return(cell)
}

# ... more code that emits raw Typst using the above function goes here...

cat("\n```")
```
````

And here's what the Typst file looks like, starting at the above code (so skipping the Quarto template):

````Typst


#set text(font: "Roboto")

#set page(
  header: align(right)[
    #image("logo.png", width: 15%) 
  ]
)

#show heading.where(level: 1): it => [
  #set text(24pt)
  #upper(it.body)
]

#show heading.where(level: 2): it => [
  #set text(18pt)
  #upper(it.body)
]

#let cell = rect.with(
  inset: 8pt,
  stroke: white,
  width: 100%,
  radius: 6pt)

#let fa(codepoint) = text(font: "Font Awesome 6 Free", codepoint)
#let calendar = fa("\u{f133}")

#let deemph(x) = text(fill: rgb("1978b0"), x)

= My title here \
As of #datetime.today().display("[day] [month repr:long] [year]")
  \
  \

#grid(
  columns: 2,
  rows: (100pt, 100pt, 100pt, 100pt),
  column-gutter: 10pt,
  row-gutter: 10pt,

cell(height: 100%)[18 Mar\
*REDACTED TITLE*\
#deemph[#calendar REDACTED CONTENT...] \
],

````

And so on with the rest of the content. It's definitely not as sophisticated as defining your own Knitr engine, but maybe it's possible for you to blend the approaches and emit any set rules directly before your other content? Not sure why it's different to your own tests with a custom knit engine — maybe `{=rawblock}` is handled separately to knitr?

**Reply by baptiste** [reply]
I think using cat() lets you bypass the wrapper, but unfortunately that's not an option with the engine output(?)

**Reply by jimjam-slam** [reply]
Could you piggyback off the [cat engine](https://bookdown.org/yihui/rmarkdown-cookbook/eng-cat.html)? They actually have an example on [syntax highlighting YAML values](https://bookdown.org/yihui/rmarkdown-cookbook/eng-cat.html#write-yaml-data-to-a-file-and-also-display-it)!

**Reply by cscheid** [reply]
If you use `output: asis` _with no other options_ and `cat()` the raw block, Quarto won't wrap it in any output:

`````
---
engine: knitr
format: typst
keep-md: true
keep-typ: true
---

````{r}
#| output: asis
cat("```{=typst}\n#set text(font: \"Roboto\")\n```")
````

`````

keep-md output:

````
---
engine: knitr
format: typst
keep-md: true
keep-typ: true
---



```{.r .cell-code}
cat("```{=typst}\n#set text(font: \"Roboto\")\n```")
```


```{=typst}
#set text(font: "Roboto")
```
````

**Reply by baptiste** [reply]
Unfortunately, I still don't see a way to use `cat()` inside a knitr engine definition. I think it's capturing the output and not showing it.
I've more or less moved on from this original idea, and I'm now contemplating a Lua filter rather than a custom knit engine, or using the cat engine with an intermediate external file, but it's a bit unfortunate that it's not easier to achieve a _raw_ output out of a custom engine chunk.

Specifically, this produces no output:

````
---
title: "Test"
format: 
  typst:
    keep-typ: true
    keep-md: true
---

```{r, echo=FALSE}
library(glue)
library(yaml)
library(knitr)

knitr::knit_engines$set(yamltypst = function(options) {

    options[["results"]] <- "asis"
    options[["echo"]] <- FALSE

    # get code together
    code <- paste(options$code, collapse = '\n')

    cat(code, file='')

    knitr::engine_output(
      options,
      options$code,
      out = NULL
    )
  })
```

## Quarto

Quarto enables you to weave together content and executable code into a finished document.

```{yamltypst}
stuff: yamlcontent
```
````

**Reply by mcanouil** [reply]
Did you try `knitr::asis_output()`?

**Reply by cderv** [reply]
> I think Quarto is wrapping your command around a Div, and then (as it stands), I think you're out of luck. Can you give us a minimal example? We have mechanisms for stripping structural Divs out of documents, and we should consider using it here.

@cscheid I do think this could be the the issue here.  Should our internal divs creates a `#block` in the output ? Or maybe we should trip the unprocessed one 

````markdown
---
title: "Test"
format: 
  typst:
    keep-typ: true
    keep-md: true
---

```{r}
1 + 1
```

````

````markdown
---
title: "Test"
format: 
  typst:
    keep-typ: true
    keep-md: true
---


::: {.cell}

```{.r .cell-code}
1 + 1
```

::: {.cell-output .cell-output-stdout}

```
[1] 2
```


:::
:::

````

You can observe this behavior with example share at https://github.com/quarto-dev/quarto-cli/discussions/9163#discussioncomment-8876551 with the engine creation. 

@baptiste you should not need to use `cat()`. If you add `output: asis` to the cell using your engine, you do get the expected results right ? 
````markdown
```{yamltypst}
#| output: asis
stuff: yamlcontent
```
````

I believe the "asis" set in the engine gets picked up, and correctly remove the div around the output of the cell
````r
options[["results"]] <- "asis"
````

but it seems it is does not prevent the external div, we should not get included if only ``::: {.cell}`. I'll look into that because this means `asis` is not get picked up completely for R cells options. 🤔 

So for now, `output: asis` on the cells itself needs to be set for Quarto to do the right things. 

Can you try that ? 


> Did you try knitr::asis_output()?

@mcanouil this won't work either unfortunately - same issue - this function sets `options$results = "asis"` for the output hook to do the right thing, but Quarto does not pick the value. When you use a custom engine, no need to use `asis_output()`. 

> Unfortunately, I still don't see a way to use cat() inside a knitr engine definition. I think it's capturing the output and not showing it.

@baptiste your initial engine function is ok to me, and using `cat()` inside custom engine is indeed not something that would work. You are correctly using `engine_output()`. This is on Quarto here not processing the knitr options the same as in R Markdown, and outputting this divs where they should not be, and our Lua process does not either strip them. 

Unlucky combination

**Reply by cderv** [reply]
FWIW I have opened two issues related to knitr behavior
* https://github.com/quarto-dev/quarto-cli/issues/9184
* https://github.com/quarto-dev/quarto-cli/issues/9185

**Reply by baptiste** [reply]
Indeed, this works:

````
---
title: "Test"
format: 
  typst:
    keep-typ: true
    keep-md: true
---

```{r, echo=FALSE}
library(glue)
library(yaml)
library(knitr)

knitr::knit_engines$set(yamltypst = function(options) {

    # options[["results"]] <- "asis"

    # get code together
    code <- paste(options$code, collapse = '\n')
    # reset language for markdown highlighting
    options$engine <- "yaml"

    out <- knitr::raw_block(glue::glue("#strong(  [{ spec }] )"), "typst")
    
    knitr::engine_output(
      options,
      options$code,
      out = out
    )
  })
```

## Quarto

Quarto enables you to weave together content and executable code into a finished document.

```{yamltypst}
#| output: asis
stuff: yamlcontent
```

````

Thanks for disentangling the thread, @cderv ; it hadn't occurred to me that setting the engine defaults was going to interfere with the processing of the chunk.

**Reply by baptiste** [reply]
By the way, just to illustrate the scenario where this came up: I'm trying to format an agenda from plain text entries, and I find Yaml the most convenient syntax for it (with all caveats that apply). It is of course possible to enter Yaml in a separate file, but I prefer to keep everything in one source document. It's also possible to enter the Yaml between quotes within a code chunk or raw typst chunk, but this get really annoying as the`"` need to be escaped inside, the text loses all syntax highlighting in the editor, and there are obscure looking commands surrounding the content that won't mean anything to someone not (deeply) familiar with quarto/knitr, if I give them the plain text.

For these reasons, I toyed with the idea of having the Yaml embedded in the document as a custom language engine, where it would appear something like this:

````
```{agenda}
"admin":
  Topic: Misc. admin
  Time: 5 mins
  Lead: baptiste 
  Purpose: Inform, Admin
  Preparation: |
    - Anything to change on the agenda?
    - Read minutes from last meeting
    - Review action points
  Process: |
    - Check if agenda needs to be adjusted
    - Check if everyone's happy with the minutes
    - List top priorities of meeting and items needing decisions
... 
```
````
and the custom engine "agenda" would produce,

````
#let ag = yaml.decode(" <contents injected here> ")
#agenda-table(ag)
````
with the Typst macro [defined elsewhere](https://typst.app/project/rX1dOa_DqSy46fiFCPTFeZ).

<img width="1371" alt="Screenshot 2024-03-26 at 09 53 29" src="https://github.com/quarto-dev/quarto-cli/assets/18970/e12dc77e-6d95-4f5b-bcf6-57ca3fde9870">

I think this kind of use-case has some appeal, and it could be nice to think of a more generic quarto interface to get this type of functionality (thinking along the lines of the glue, cat, knitdata kinds of engine in knitr, but with the purpose of injecting the block's raw content into a custom string).

**Reply by cscheid** [reply]
This is really nice, yeah!

I wonder what it would look like if we tried to take advantage of `{{< meta >}}` or maybe even just a filter. The main issue is that we'd need to make YAML metadata conveniently available to execution engines.

With a plain Lua filter, you could try something like the following (your tolerance for Lua allowing for it, of course):

````
---
format: html
filters:
  - agenda.lua
---

Now, for the agenda:

---
agenda: 
  "admin":
    Topic: Misc. admin
    Time: 5 mins
    Lead: baptiste 
    Purpose: Inform, Admin
    Preparation: |
      - Anything to change on the agenda?
      - Read minutes from last meeting
      - Review action points
    Process: |
      - Check if agenda needs to be adjusted
      - Check if everyone's happy with the minutes
      - List top priorities of meeting and items needing decisions
  "snoopy update":
    ...
---

::: agenda

- `admin`
- `snoopy update`

:::
````

And then

```lua
-- agenda.lua

function Meta(meta)
  entries = meta['agenda'] or {}
end

function Div(div)
  if not div.classes:includes('agenda') then
    return nil
  end
  local entries = pandoc.List({})
  div:walk({
    Code = function(code)
      entries:insert(...) -- do your thing here
    end
  })
  -- do your thing here
  return result
end
```

**Reply by baptiste** [reply]
Nice! Indeed, the lua filter approach is a good alternative because it won't rely on knitr. 

I have something more generic in mind though, because here we're relying on the content being Yaml, which happens to coincide with quarto+pandoc, but it's easy to imagine similar input-wrangling activities with other plain text formats, such as TOML, JSON, or even the likes of Mermaid, TIkz, etc. where, unlike the current workflows, the raw content would be injected into the document's output format rather than evaluated. 
Imagine a Typst library that would convert Tikz code on-the-fly into its native Cetz library, for example, or draw diagrams from a Graphviz input representation, etc. – it'd be nice to have the quarto input written in a simple cell (not sure what format is most suitable: ```` ```{=custom} ```` or `` ::: custom ``, or ```` ```{custom} ````etc.) and have a Lua filter grab the content and inject it in a user-specified string. For maximum flexibility this would be a user-defined template a la glue, say:

````
```{agenda}
#| highlight : yaml
#| wrap : "#let a = yaml.decode( \"{agenda-code}\" )"
...
```
````
where those arguments could also be set as defaults for the whole document.

**Reply by cscheid** [reply]
> a Lua filter grab the content and inject it in a user-specified string

I don't like this approach because it's very hard to make it work well, especially with the myriad escaping problems that can come up. If you're going to do something like that, you really ought to be sending a very boring string, like a base64-encoded JSON blob. Imagine having to deal with escaping single vs double vs smart quotes, escaping all different possible new line characters, etc.

**Reply by baptiste** [reply]
True, and i did run into this, but insofar as typst is concerned there seems to be a good workaround, which is to wrap the content as raw, and then retrieve the text with .text :

```
#let stuff = `{inject here)`
#yaml.decode(stuff.text)
```

**Reply by baptiste** [reply]
It's now a typst template: https://typst.app/universe/package/yagenda

I'm still pondering the best approach to use it in quarto; [here's the custom knit-engine version](https://gist.github.com/baptiste/13214e4cc77a7143fb7be2d5ac1a5a20) (requires using a more recent Typst to compile, I set it with `export QUARTO_TYPST=/usr/local/bin/typst` at my own risks) 

In practice I'm finding that having the YAML in an external file is actually alright, but I still like the idea of being able to embedded it in the qmd file directly, especially when it's very short. Currently the chunk options are a little obtrusive, and obviously the first setup chunk is very distracting.

**Reply by cderv** [reply]
> requires using a more recent Typst to compile

Which version do you need ? Our current 1.5 pre-release now includes Typst 0.11.0 (since 1.5.29). 

> Currently the chunk options are a little obtrusive, and obviously the first setup chunk is very distracting.

You would need a companion R package that registers this engine on load or through a function call. This would avoid this long setup chunk. (Sourcing a R script for a URL like a gist is also possible but not really a good practice I guess)

> In practice I'm finding that having the YAML in an external file is actually alright

One think that could work with **knitr** built in engine would be to write to file the content of the chunk, so that you can load it in your raw typst block. 

````markdown
---
name: "Pumpkins and peanuts committee"
date: "03/01/2000"
time: "2 pm"
location: "baseball field"
invited: 
  - Sally
  - Shroeder
  - Pig-pen
  - Marcie
format: 
  typst:
    keep-typ: true
    keep-md: true
    template-partials:
      - typst-show.typ
        
engine: knitr
---

```{cat}
#| engine-opts:
#|   file: agenda.yaml  
#| echo: false
"admin":
  Topic: Misc. admin
  Time: 5 mins
  Lead: baptiste 
  Purpose: Inform, Admin
  Preparation: |
    - Anything to change on the agenda? 
    - Read minutes from last meeting
    - Review action points
  Process: |
    - Check if agenda needs to be adjusted
    - Check if everyone's happy with the minutes
    - List top priorities of meeting and items needing decisions
    
"snoopy update":
  Topic: Snoopy's latest 
  Time: 10 mins
  Lead: Charlie Brown
  Purpose: Share, Discuss
  Preparation: |
    - Bring your favorite comic strip
    - Reflect on Snoopy's recent escapades
  Process: |
    - Discuss Snoopy's ongoing "battle" with the Red Baron
    - Share anecdotes of Snoopy's fiercest war moves
    - Plan a group outing to visit Snoopy's doghouse
"woodstock sighting":
  Topic: Woodstock's whereabouts
  Time: 5 mins
  Lead: Linus
  Purpose: Locate, Update
  Preparation: |
    - Bring binoculars for bird-watching
    - Brush up on bird calls
  Process: |
    - Share any recent sightings or chirpings from Woodstock
    - Discuss strategies for bringing Woodstock back to the tunes
    - Plan an upside-down birdhouse-building workshop 
"great pumpkin plans":
  Topic: Prep for Great Pumpkin
  Time: 15 mins
  Lead: Linus
  Purpose: Plan, Excite
  Preparation: |
    - Bring your pumpkin-carving tools
    - Practice your most sincere pumpkin patch speech
  Process: |
    - Discuss tactics for maximizing sincerity in the pumpkin patch
    - Brainstorm new ways to attract the Great Pumpkin's attention
    - Assign roles for the annual pumpkin carving contest
"philosophical discussion":
  Topic: Philosophical Musings
  Time: 20 mins
  Lead: Lucy and Charles
  Purpose: Ponder, Reflect
  Preparation: |
    - Bring your favorite existentialist quotes
    - Contemplate the meaning of life #v(100%)
  Process: |
    - Engage in deep philosophical discussions under the night sky
    - Debate the nature of happiness, existence, and 5c psychiatry
    - Seek wisdom from Linus's trusty security blanket
"beethoven appreciation":
  Topic: Beethoven's Legacy
  Time: 10 mins
  Lead: Schroeder
  Purpose: Appreciate, Discuss
  Preparation: |
    - Bring your favorite Beethoven compositions
    - Practice your air piano skills
  Process: |
    - Listen to Schroeder perform select Beethoven pieces
    - Discuss the timeless appeal of Beethoven's music
    - Plan a Beethoven-themed recital for Christmas
"baseball game":
  Topic: Baseball Game
  Time: 30 mins
  Lead: Charlie Brown
  Purpose: Play, Bond
  Preparation: |
    - Bring your baseball glove and bat
    - Review the rules of baseball
  Process: |
    - Divide into teams for an exciting game of baseball
    - Cheer on Charlie Brown as he attempts to finally kick that football
    - Enjoy snacks and camaraderie under the shade of the old oak tree
    
"summer camp":
  Topic: Summer Camp Adventures
  Time: 15 mins
  Lead: Peppermint Patty
  Purpose: Plan, Excite
  Preparation: |
    - Bring ideas for summer camp activities!
    - Check availability of camping gear!
  Process: |
    - Brainstorm camping trip destinations and outdoor activities
    - Discuss potential guest speakers or counselors
    - Organize a talent show and marshmallow roasting competition
"marcie school update":
  Topic: Marcie's Academic Progress
  Time: 10 mins
  Lead: Peppermint Patty
  Purpose: Discuss, Support
  Preparation: |
    - Bring Marcie's recent report card
    - Reflect on Marcie's study habits and strengths
  Process: |
    - Share updates on Marcie's academic achievements and challenges
    - Discuss strategies to help Marcie excel in school
    - Offer encouragement and support to Marcie in her studies
"linus blanket workshop":
  Topic: Linus and his blanket
  Time: 20 mins
  Lead: Lucy
  Purpose: Understand, *Support*
  Preparation: |
    - Bring your own security blanket (optional)
    - Reflect on the significance of Linus's blanket
  Process: |
    - Discuss the history and symbolism of Linus's security blanket
    - Explore ways to help both Linus and Snoopy 
    - Release a blanket statement
"snoopy plane fights":
  Topic: Snoopy vs the Baron
  Time: 10 mins
  Lead: Woodstock
  Purpose: Support
  Preparation: |
    - Bring your favorite Snoopy flying ace memory
    - Reflect on Snoopy's aerial prowess
  Process: |
    - Share anecdotes from Snoopy's battles
    - Discuss the psychological implications of Snoopy's dogging of bullets
    
```

```{=typst}
#let yamlentries = yaml("agenda.yaml")
#agenda-table(yamlentries)
```

````

The raw block could even be set in a template partials directly, or a header include directly. 

If you package all this in a Quarto extension, you could just get an agenda output using 
- A `cat` knitr engine to write to YAML file
- a `agenda.yaml` in the project document root

The extension could have a shortcode to place the agenda in the right place in the document body. 

Using Lua approach discussed above is also a good one to avoid depending on a computation engine to write YAML from a Markdown formatted content (YAML meta block or special div block, or special raw block with attributes). You could imagine do the same logic (get yaml from the in document and write to a file or inset directly) from a code block 

````
```{.yaml .agenda}
...
```
````
where you could catch this code block, and process it for your agenda template. 

Anyhow, great Typst template!

**Comment by baptiste** [comment]
> Which version do you need ? Our current 1.5 pre-release now includes Typst 0.11.0 (since 1.5.29).

Thanks, I hadn't seen that recent change. 

> > Currently the chunk options are a little obtrusive, and obviously the first setup chunk is very distracting.
> 
> You would need a companion R package that registers this engine on load or through a function call. This would avoid this long setup chunk. (Sourcing a R script for a URL like a gist is also possible but not really a good practice I guess)

True, I was mostly aiming for a proof of principle so I didn't wrap it in a package. I wish there was a way to specify a "setup" file in the YAML header, if just to load a set of  packages. Something like 

```
knitr:
   startup: 
      text: |
         library(glue)
         source("fun.R")
```

> > In practice I'm finding that having the YAML in an external file is actually alright
> 
> One think that could work with **knitr** built in engine would be to write to file the content of the chunk, so that you can load it in your raw typst block.
> ...
> If you package all this in a Quarto extension, you could just get an agenda output using
> 
>     * A `cat` knitr engine to write to YAML file
> 
>     * a `agenda.yaml` in the project document root
> 
> The extension could have a shortcode to place the agenda in the right place in the document body.

Yep, cat() is the route I initially took (without wrapping it as an extension), and I had the filename specified in the YAML for more flexibility:

````
---
agenda: "2024-04-05-agenda.yaml"
---

```{r, echo=FALSE, results='hide'}
meta <- rmarkdown::metadata
agenda <- meta$agenda
```

````

> Using Lua approach discussed above is also a good one to avoid depending on a computation engine to write YAML from a Markdown formatted content (YAML meta block or special div block, or special raw block with attributes). You could imagine do the same logic (get yaml from the in document and write to a file or inset directly) from a code block
> 
> ```
> ```{.yaml .agenda}
> ...
> ```
> ```
> 
> where you could catch this code block, and process it for your agenda template.

Yes, I think this is the best option, so I'm hoping to try it at some point. Unfortunately the Lua side of things is always a bit of a barrier :/
I imagine the logic could be:

````
function Code(code)
  if not code.classes:includes('agenda') then
    return nil
  end
  -- capture the code block as raw text
  local rawcode = code.content ???
  
  -- append some typst wrapper
  local fullstring = "#let yamlcontent = ` " + rawcode + " `.text "
  -- or write the content in an external file like the cat engine does
  pandoc.write(rawcode, filename)
  
  -- output raw typst block
  pandoc.raw(fullstring, "typst")

  return result
end
````


> Anyhow, great Typst template!

Thanks! Credits should go to PgSuper who wrote the table code; I merely tweaked it and wrapped it up.

**Reply by cderv** [reply]
> I wish there was a way to specify a "setup" file in the YAML header, if just to load a set of packages. Something like

yeah there is no such thing in **knitr** - you need to include as code cell what you want to be run in the knitting process. 

I guess the thinking was that it is always easy enough to add the code as startup chunk 
````markdown
```{r}
#| label: startup
library(glue)
source(fun.R)
```
````

Potentiall in a share included file to be added in all document. 

Startup file like `.Rprofile` is run by R though - I don't think we run with `--vanilla` but there will be the topic of R working directory at startup to pick up a project .Rprofile. So not  really safe probably. 

This topic has come up several time, especially because Quarto starts new process for engine each time and you may want some common configuration for all your runs. Like startup file but tied to quarto execution. 

Interesting feedback thanks ! 

> Unfortunately the Lua side of things is always a bit of a barrier :/ I imagine the logic could be:

Yes Lua requires a bit of a learning curve, but we hope our targetted documentation and existing filters help find example to start with. 
https://quarto.org/docs/extensions/lua.html

Some comment in the logic to give you hint (that you could have found yourself probably 😅 )

- `CodeBlock` would be the element to search for - see pandoc's doc for its content https://pandoc.org/lua-filters.html#type-codeblock
- `code.text` would have the content of the block
- `pandoc.RawBlock()` would be what to use for creating some Raw Content from AST for pandoc to write asis
- `pandoc.write()` is not a writing to file function, this is to run a Pandoc writer to convert. See io and file base API for writing to file (https://quarto.org/docs/extensions/lua-api.html#lua-base-api) 
- `..` is what to use to concatenate (not `+`)

THe logic is the right one, and you are really close with using the right syntax ! Good little exercise :)

**Reply by baptiste** [reply]
> THe logic is the right one, and you are really close with using the right syntax ! Good little exercise :)

Full disclosure, I did not even try to check any of the syntax when I wrote this at breakfast :P For me the barrier with Lua is more of a debugging one: I have not found a setup where I can just try something and iterate like in R's REPL, but with the quarto+pandoc environments and functions at hand like the quarto process has them set up. Haven't tried for a while though, so I should definitely revisit the docs.

**Reply by cderv** [reply]
>  I have not found a setup where I can just try something and iterate like in R's REPL, but with the quarto+pandoc environments and functions at hand like the quarto process has them set up

Yeah I can relate to that. I don't have anything great to offer. I am mainly relying of old good console log debugging. So I use `quarto.log.output()` a lot to print where I want to check and debug things. 

For non specific Quarto lua processing , `pandoc lua -i` has been also helpful to run Lua interactively in console (available since 3.1.2 https://pandoc.org/releases.html#pandoc-3.1.2-2023-03-26). It works only in bare pandoc context for now. 

I never tried setting debugging with Quarto directly : https://pandoc.org/lua-filters.html#debugging-lua-filters

Hope it helps
